bundles/org.openhab.binding.verisure/src/main/java/org/openhab/binding/verisure/internal/dto/VerisureEventLogDTO.java

```diff
@@ -0,0 +1,294 @@
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.verisure.internal.dto;
+
+import static org.openhab.binding.verisure.internal.VerisureBindingConstants.THING_TYPE_EVENT_LOG;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.commons.lang.builder.EqualsBuilder;
+import org.apache.commons.lang.builder.HashCodeBuilder;
+import org.apache.commons.lang.builder.ToStringBuilder;
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+
+import com.google.gson.annotations.SerializedName;
+
+/**
+ * The event log of the Verisure System.
+ *
+ * @author Jan Gustafsson - Initial contribution
+ *
+ */
+@NonNullByDefault
+public class VerisureEventLogDTO extends VerisureBaseThingDTO {
+
+    private Data data = new Data();
+
+    public Data getData() {
+        return data;
+    }
+
+    public void setData(Data data) {
+        this.data = data;
+    }
+
+    @Override
+    public ThingTypeUID getThingTypeUID() {
+        return THING_TYPE_EVENT_LOG;
+    }
+
+    @Override
+    public String toString() {
+        return new ToStringBuilder(this).append("data", data).toString();
+    }
+
+    @Override
+    public int hashCode() {
+        return new HashCodeBuilder().append(data).toHashCode();
+    }
+
+    @Override
+    public boolean equals(@Nullable Object other) {
+        if (other == this) {
+            return true;
+        }
+        if (!(other instanceof VerisureEventLogDTO)) {
+            return false;
+        }
+        VerisureEventLogDTO rhs = ((VerisureEventLogDTO) other);
+        return new EqualsBuilder().append(data, rhs.data).isEquals();
+    }
+
+    public class Data {
+
+        private Installation installation = new Installation();
+
+        public Installation getInstallation() {
+            return installation;
+        }
+
+        @Override
+        public String toString() {
+            return new ToStringBuilder(this).append("installation", installation).toString();
+        }
+
+        @Override
+        public int hashCode() {
+            return new HashCodeBuilder().append(installation).toHashCode();
+        }
+
+        @Override
+        public boolean equals(@Nullable Object other) {
+            if (other == this) {
+                return true;
+            }
+            if (!(other instanceof Data)) {
+                return false;
+            }
+            Data rhs = ((Data) other);
+            return new EqualsBuilder().append(installation, rhs.installation).isEquals();
+        }
+    }
+
+    public class Installation {
+
+        private EventLog eventLog = new EventLog();
+        @SerializedName("__typename")
+        private @Nullable String typename;
```

Comment:
I think it would be safe to move Data and Installation into their own separate class that can be shared among all of the DTO classes.

You should be able to have a single `Installation` class that combines all the fields from all the DTO Installation classes. Json parsing won't be an issue since it simply won't populate the fields that weren't present in the json result. And the code that uses each DTO object already knows what kind of fields to expect so you can ignore the fields it doesn't care about.

I suspect you could save a thousand lines of code by doing this.

Reply:
I'll look into this.

Reply:
I've moved most of the DTO classes to VerisureBaseThingDTO and added all needed fields to them.